File: R/read-thresholds.R
# Reader for the optional per-kit LOD/LLOQ/ULOQ threshold xlsx files
# (DESIGN.md §2.5). The thresholds are not used to recompute statuses;
# the WebIDQ export already carries authoritative status annotations.

# Pull `KIT\d+` out of a file's basename. Used because the threshold xlsx
# itself does not carry the kit ID anywhere (decisions.md #51).
.kit_id_from_filename <- function(path) {
  base <- basename(path)
  pos <- regexpr("KIT\\d+", base)
  if (is.na(pos) || pos < 1L) {
    stop("Could not find a `KIT\\d+` token in the threshold file name '",
         base, "'. Rename the file to include the kit ID (e.g. ",
         "`thresholds_KIT01.xlsx`).", call. = FALSE)
  }
  regmatches(base, pos)
}

# Read one per-kit threshold xlsx into a list with the four fields needed
# to assemble a ThresholdData. Vendor exports (decisions.md #51) use a
# 4 rows × (1 + N) cols layout — sheet name is not validated:
#   row 1: "Measurement time" + metabolite names
#   row 2: "LOD (calc.)" + LOD values
#   row 3: "ULOQ"        + ULOQ values
#   row 4: "LLOQ"        + LLOQ values
# Kit ID is recovered from the filename, not from any cell.
.read_one_threshold_file <- function(path) {
  raw <- readxl::read_excel(path, col_names = FALSE,
                            .name_repair = "minimal")
  raw <- as.matrix(as.data.frame(raw, stringsAsFactors = FALSE,
                                 check.names = FALSE))
  if (nrow(raw) < 4L) {
    stop("Threshold file '", path, "' has fewer than 4 rows; expected ",
         "Measurement-time / LOD / ULOQ / LLOQ rows.", call. = FALSE)
  }
  kit_id <- .kit_id_from_filename(path)
  metabolite_names <- as.character(raw[1L, -1L, drop = TRUE])
  lod  <- suppressWarnings(as.numeric(raw[2L, -1L, drop = TRUE]))
  uloq <- suppressWarnings(as.numeric(raw[3L, -1L, drop = TRUE]))
  lloq <- suppressWarnings(as.numeric(raw[4L, -1L, drop = TRUE]))
  list(kit_id = kit_id,
       lod = lod, lloq = lloq, uloq = uloq,
       metabolite_names = metabolite_names)
}

#' Read per-kit LOD/LLOQ/ULOQ threshold files
#'
#' Optional metadata reader. See DESIGN.md §2.5. Each input file covers one
#' kit; supply one path per kit and they are stacked by kit. The
#' thresholds are not used to recompute statuses (that information comes
#' from the WebIDQ export); they are attached to the assay row metadata so
#' downstream reports can cite them.
#'
#' @param threshold_files Character vector of paths to per-kit threshold
#'   xlsx files (e.g., `c("lod_kit01.xlsx", "lod_kit02.xlsx")`).
#' @param synonyms Optional named character vector mapping typed metabolite
#'   names in the threshold file to canonical short names.
#' @return A [ThresholdData] object.
#' @export
read_thresholds <- function(threshold_files, synonyms = NULL) {
  if (length(threshold_files) == 0L) {
    stop("threshold_files must contain at least one path.", call. = FALSE)
  }
  per_kit <- lapply(threshold_files, .read_one_threshold_file)

  ref_metabs <- per_kit[[1L]]$metabolite_names
  for (i in seq_along(per_kit)) {
    if (!identical(per_kit[[i]]$metabolite_names, ref_metabs)) {
      stop("Threshold files do not share the same metabolite columns.",
           call. = FALSE)
    }
  }
  resolved <- .resolve_metabolite_names(ref_metabs, synonyms)
  kit_ids <- vapply(per_kit, `[[`, character(1L), "kit_id")

  to_matrix <- function(field) {
    m <- do.call(rbind, lapply(per_kit, `[[`, field))
    rownames(m) <- kit_ids
    colnames(m) <- resolved
    m
  }

  ThresholdData(
    lod = to_matrix("lod"),
    lloq = to_matrix("lloq"),
    uloq = to_matrix("uloq"),
    kit_ids = kit_ids,
    metabolite_names = resolved
  )
}

File: python/metabosetr/read_thresholds.py
"""Optional per-kit LOD/LLOQ/ULOQ reader (port of ``read-thresholds.R``).

DESIGN.md §2.5. Thresholds are metadata only -- they are not used to
recompute statuses. Kit ID comes from the filename (``KIT\\d+``); the file
itself does not carry it.
"""

from __future__ import annotations

import re
from pathlib import Path
from typing import Optional

import numpy as np

from . import _xlsx
from .classes import ThresholdData
from .read_webidq import _resolve_metabolite_names

_KIT_RE = re.compile(r"KIT\d+")


def _kit_id_from_filename(path: str) -> str:
    base = Path(path).name
    m = _KIT_RE.search(base)
    if not m:
        raise ValueError(
            f"Could not find a `KIT\\d+` token in the threshold file name "
            f"'{base}'. Rename the file to include the kit ID "
            "(e.g. `thresholds_KIT01.xlsx`)."
        )
    return m.group(0)


def _to_float_row(cells) -> np.ndarray:
    out = np.empty(len(cells), dtype=float)
    for i, v in enumerate(cells):
        try:
            out[i] = float(v)
        except (TypeError, ValueError):
            out[i] = np.nan
    return out


def _read_one_threshold_file(path: str) -> dict:
    """Read one per-kit threshold sheet (4 x (1+N) layout)."""
    _, rows = _xlsx.first_sheet_rows(path)
    if len(rows) < 4:
        raise ValueError(
            f"Threshold file '{path}' has fewer than 4 rows; expected "
            "Measurement-time / LOD / ULOQ / LLOQ rows."
        )
    metabolite_names = [str(v) for v in rows[0][1:]]
    lod = _to_float_row(rows[1][1:])
    uloq = _to_float_row(rows[2][1:])
    lloq = _to_float_row(rows[3][1:])
    return {
        "kit_id": _kit_id_from_filename(path),
        "lod": lod,
        "uloq": uloq,
        "lloq": lloq,
        "metabolite_names": metabolite_names,
    }


def read_thresholds(threshold_files, synonyms: Optional[dict] = None) -> ThresholdData:
    """Read per-kit LOD/LLOQ/ULOQ files into a :class:`ThresholdData`.

    One file per kit; supply one path per kit and they are stacked by kit.
    """
    if isinstance(threshold_files, str):
        threshold_files = [threshold_files]
    if len(threshold_files) == 0:
        raise ValueError("threshold_files must contain at least one path.")

    per_kit = [_read_one_threshold_file(p) for p in threshold_files]

    ref_metabs = per_kit[0]["metabolite_names"]
    for pk in per_kit:
        if pk["metabolite_names"] != ref_metabs:
            raise ValueError("Threshold files do not share the same metabolite columns.")

    resolved = _resolve_metabolite_names(ref_metabs, synonyms)
    kit_ids = [pk["kit_id"] for pk in per_kit]

    def to_matrix(field: str) -> np.ndarray:
        return np.vstack([pk[field] for pk in per_kit])

    return ThresholdData(
        lod=to_matrix("lod"),
        lloq=to_matrix("lloq"),
        uloq=to_matrix("uloq"),
        kit_ids=kit_ids,
        metabolite_names=resolved,
    )
